neqo-transport/src/packet/mod.rs

```diff
@@ -304,6 +304,12 @@ impl PacketBuilder {
 
     /// Build the packet and return the encoder.
     pub fn build(mut self, crypto: &mut CryptoDxState) -> Res<Encoder> {
+        if self.len() > self.limit {
```

Comment:
I think we should put a similar check and debug_assert! in every place we write into the buffer. If anyone catch the assert we will know exactly where it is happening. Also add a message to the assertion with useful values, e.g. self.len() and self.limit.
Also if we want to catch it sooner we could add  a different message to every place we close with InternalError and then we collect telemetry about it (that was my idea how to debug #1052 ). We can collect that telemetry on Nightly for a couple of weeks to catch all internal error or the most frequent errors.